lib/api/Healthcheck.js

```diff
@@ -101,6 +115,174 @@ class Healthcheck {
             }
         });
     }
+
+    /**
+     * Get deep healthcheck
+     * A deep healthcheck will send entries to BackbeatProducer to produce to
+     * Kafka, catch these entries using the same global Kafka topic used for
+     * Backbeat CRR, but using a new BackbeatConsumer for each health request
+     * @param {function} cb - callback(error, data)
+     * @return {undefined}
+     */
+    getDeepHealthcheck(cb) {
+        if (this._probe !== null) {
+            // TODO: Based on previous request, given how deep healthcheck
+            //  currently works, cannot handle parallel requests above cluster
+            //  count. Will need to refactor this
+            const error = {
+                method: 'Healthcheck.getDeepHealthcheck',
+                error: 'Another request still processing',
+            };
+            return cb(error);
+        }
+        if (!this._crrProducer || !this._crrStatusProducer) {
+            const error = {
+                method: 'Healthcheck.getDeepHealthcheck',
+                error: 'Backbeat Producer not set',
+            };
+            return cb(error);
+        }
+
+        const start = Date.now();
+        const id = `deephealthcheck-${uuid().replace(/-/g, '')}`;
+        this._probe = new Probe(id);
+
+        return async.waterfall([
+            done => this._getTopicPartitions(done),
+            (partitions, done) => {
+                this._probe.setupStore(partitions);
+                this._createEntries(id, partitions, done);
+            },
+            (entries, done) => this._crrProducer.send(entries, done),
+            done => this._createConsumer('deephealthcheck-group', done),
+        ], (err, consumer) => {
+            if (err || consumer === undefined) {
+                const error = {
+                    method: 'Healthcheck.getDeepHealthcheck',
+                    error: err,
+                };
+                return cb(error);
+            }
+            this._probe.on('collect', (partition, id) => {
+                if (id === this._probe.getId()) {
+                    this._probe.setStoreData(partition, 'ok');
+                }
+            });
+
+            return async.retry({
+                times: Number.parseInt(DH_TIMELIMIT / DH_INTERVAL, 10),
+                interval: DH_INTERVAL,
+                errorFilter: err => err === 'retry',
+            },
+            this._waitAndCheck.bind(this, start),
+            (err, elapsed) => {
+                if (err) {
+                    // replace all undefined values as 'error'
+                    this._probe.setStoreErrors();
+                }
+                this._probe.setStoreData('timeElapsed', elapsed);
```

Comment:
Looks like this needs to be in an `else` block.

Reply:
I want to set this property no matter success or failure. Just thought to add a time elapsed indicator since the time it takes for a request to finish is unknown 👍